### PR TITLE
Make runtime api calls native when possible

### PR DESCRIPTION
**Pull request by bkchr**
This pull request implements support for making real native calls, without using the dispatch function in the native runtime.
The dispatch function is not removed, because for stuff like RPC it is still used.

The native call also does not drop encode/decode completely. For parameters/return values that are generic over the `Block`, will be encoded/decoded before calling the native function. This is a requirement, as the block differs between the node and the runtime.

There are still some optimizations, like don't encode all parameters, even if we only call the native function. Currently this needs to be done, as the state machine decides if it calls both wasm/native or just native or just wasm.  

Fixes: https://github.com/paritytech/substrate/issues/294

### CODE REVIEW

**Comment by gavofyork** [comment]
@arkpar could you take a look?

File: core/client/src/block_builder/block_builder.rs [hunk 1]
```diff
@@ -89,7 +89,7 @@ where
 			extrinsics: &mut Vec<Block::Extrinsic>
 		) -> error::Result<()> where T: BlockBuilderApi<Block, InherentData> {
 			api.map_api_result(|api| {
-				match api.apply_extrinsic(block_id, &xt)? {
+				match api.apply_extrinsic(block_id, xt.clone())? {
```
[the remark this answers is not in the document]
Comment (bkchr) [reply core/client/src/block_builder/block_builder.rs]: Unfortunately, the implementation of `apply_extrinsics` requires to get the extrinsics by value. If we could change the implementation of `apply_extrinsics` to take them by reference, we could omit the clone here.

File: core/client/src/call_executor.rs [hunk 2]
```diff
@@ -77,7 +82,8 @@ where
 		initialised_block: &mut Option<BlockId<B>>,
 		prepare_environment_block: PB,
 		manager: ExecutionManager<EM>,
-	) -> error::Result<Vec<u8>> where ExecutionManager<EM>: Clone;
+		native_call: Option<NC>,
+	) -> error::Result<NativeOrEncoded<R>> where ExecutionManager<EM>: Clone;
```
[the remark this answers is not in the document]
Comment (bkchr) [reply core/client/src/call_executor.rs]: If we call first into native and then into wasm, we need both. I already thought about it. I did not came to a quick solution. I think in another iteration of this code, we could provide some interface, to just create `call_data`, when it is really needed.

**Comment by gavofyork** [comment]
@arkpar any other grumbles?

**Comment by bkchr** [comment]
For all that came here because they need to update their code:
A runtime api trait is not requiring to take all arguments by reference on the client side. You can now choose to take by value or by reference. This is important, as the native runtime is now able to be called, without encoding/decoding arguments.

**Pull request merged**